streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractKStreamTimeWindowAggregateProcessor.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    protected void maybeForwardUpdate(final Record<KIn, VIn> record,
+                                      final Window window,
+                                      final VAgg oldAgg,
+                                      final VAgg newAgg,
+                                      final long newTimestamp) {
+        if (emitStrategy.type() == StrategyType.ON_WINDOW_CLOSE) {
+            return;
+        } else if (tupleForwarder == null) {
+            throw new IllegalStateException("Emit strategy type is " + emitStrategy.type() + " but flush listener is not initialized.");
+        }
+
+        tupleForwarder.maybeForward(
+            record.withKey(new Windowed<>(record.key(), window))
+                .withValue(new Change<>(newAgg, sendOldValues ? oldAgg : null))
+                .withTimestamp(newTimestamp));
+    }
+
+    protected boolean shouldEmitFinal(final long closeTime) {
+        if (emitStrategy.type() != StrategyType.ON_WINDOW_CLOSE) {
+            return false;
+        }
+
+        final long now = internalProcessorContext.currentSystemTimeMs();
+        // Throttle emit frequency
+        if (now < timeTracker.nextTimeToEmit) {
+            return false;
+        }
+
+        // Schedule next emit time based on now to avoid the case that if system time jumps a lot,
+        // this can be triggered every time
+        timeTracker.nextTimeToEmit = now;
+        timeTracker.advanceNextTimeToEmit();
+
+        // Only EMIT if the window close time does progress
+        return lastEmitWindowCloseTime == ConsumerRecord.NO_TIMESTAMP || lastEmitWindowCloseTime < closeTime;
+    }
+
+    protected void fetchAndEmit(final Record<KIn, VIn> record,
+                                final long closeTime,
+                                final long emitRangeLowerBoundInclusive,
+                                final long emitRangeUpperBoundInclusive) {
+        final long startMs = time.milliseconds();
+
+        final KeyValueIterator<Windowed<KIn>, ValueAndTimestamp<VAgg>> windowToEmit = windowStore
+            .fetchAll(emitRangeLowerBoundInclusive, emitRangeUpperBoundInclusive);
+
+        int emittedCount = 0;
+        while (windowToEmit.hasNext()) {
+            emittedCount++;
+            final KeyValue<Windowed<KIn>, ValueAndTimestamp<VAgg>> kv = windowToEmit.next();
+            tupleForwarder.maybeForward(
+                record.withKey(kv.key)
+                    .withValue(new Change<>(kv.value.value(), null))
+                    .withTimestamp(kv.value.timestamp())
+                    .withHeaders(record.headers()));
+        }
+        emittedRecordsSensor.record(emittedCount);
+        emitFinalLatencySensor.record(time.milliseconds() - startMs);
+
+        lastEmitWindowCloseTime = closeTime;
+        internalProcessorContext.addProcessorMetadataKeyValue(storeName, closeTime);
+    }
+
+    abstract protected void maybeForwardFinalResult(final Record<KIn, VIn> record, final long windowCloseTime);
```

Comment:
I though `fetchAndEmit` would take care of "emit final" -- what is this method about?

Reply:
The instantiated `maybeForwardFinalResult` for time / sliding windows has a slight different validation check before calling `fetchAndEmit`.